### PR TITLE
Add all corrupted classes and methods and closing all issues

### PR DESCRIPTION
**Pull request by mEmadEl-Dien**
As we agreed, these are all files(classes and methods)

### CODE REVIEW

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp:96]: I see that this one misses its implementation.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp:144]: I think we will not work with IO space at all right? 
So why we have this here?

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/pcie_capability.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/pcie_capability.cpp:25]: I see here it misses the implementation

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_visitor_classes/configuration_write_visitor.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_visitor_classes/configuration_write_visitor.cpp:100]: this needs to be implemented in order to merge it.
So, try to finish the Capability structure first

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:13]: I understood that the TLP Constructor is something related to constructing the completion TLPs, so can you try to rename it to reduce ambiguity?

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:27]: Okay, I see that this part also depends on something that needs to be finished first

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:41]: Okay I see here you wrote that this number depends on the same thing we mentioned in another comment, so when we finish it all the other comments should be resolved.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:66]: Can you tell me why you convert from Bitset to unsingned int ?

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser.vcxproj
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser.vcxproj:16]: Actually this one which made conflict earlier, so we need to take care of it when merging.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp:11]: I know, this will be implemented, but till then this will not be resolved as finished of course.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp:23]: This one should be implemented

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp:35]: This one as well should be implemented before merging

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp:96]: Well, here I forgot to remove the comment on that line and delete the line below it (return 0), 
Also, the implementation of this function is that simple (only one line of code) but it's very general as I call it during the handling of TLP and give it a pointer to the class which I want its method to be used.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp:144]: I know we don't deal with IO (till now), but it's a part of the configuration space, so you may not implement this part, but it's okay to have it even if we don't use it.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/pcie_capability.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/pcie_capability.cpp:25]: Same the above comment regarding the same situation.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_visitor_classes/configuration_write_visitor.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_visitor_classes/configuration_write_visitor.cpp:100]: Okay I've commented out all functions regarding this class

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:13]: Indeed, it's not the TLP itself (as you mentioned), but it's the class responsible for constructing the Completion TLPs, so I can't find another name for the class, but you can suggest a name.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:27]: It just needed an explicit casting and I solved this.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:41]: Till now before implementing the PCIE capability structure it will work fine.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:66]: Cause I'm my Registers take their values as unsigned int, so when they changed their minds (Layer's team) and decided to make the data payload as bit_set, I had already made the logic for writing into the Registers, and it's not that easy that I take their payload and put it into the Register (see the function visitConfigurationSpace() which is implemented by the ConfigurationWriteVisitor for better understanding)

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser.vcxproj
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser.vcxproj:16]: You can resolve this conflict by changing this version, and the next branches (not commits in this branch) will be with version 10.0.19041.0.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp:11]: Although, it's not finished, it can't be removed as it's called inside other functions and for the project to build, and it's right to hold this all branch for these three functions which are blocked on some other function in the Layers part.
so you can merge and it has another story on it's own, and will have a branch to implement it on.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp:23]: same as above

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp
Comment (mEmadEl-Dien) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp:35]: same as above

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:13]: You Can name it as CompletionTLPConstructor maybe

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp:143]: okay 👍

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp]: okay 👍

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp]: okay cool 👍

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:66]: okay I see 👍 
nice work

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp:11]: okay no problem if they will not make future problem in something else

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp:23]: okay, I just said that, because we should merge when we finish all the functions in that specific branch, but for this case np as we will need this branch to be merged to  be able to use it

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_algorithm_classes/completion_construction_algorithms.cpp:35]: okay 👍

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser.vcxproj
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser.vcxproj:16]: yes I already solved it previously so if it makes a problem we can solve it

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:41]: okay np 👍 
just keep it in mind and not to forget it 
that 's why we don't merge un implemented functions normally

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_visitor_classes/configuration_write_visitor.cpp
Comment (david-wagih) [reply PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_visitor_classes/configuration_write_visitor.cpp:101]: okay i see 👍

**Review by david-wagih — APPROVED** [review]
Very nice work Mohamed 👍 
Keep it up 💯

**Pull request merged**
